src/Components/Components/src/RenderTree/Renderer.cs

```diff
@@ -92,6 +95,10 @@ public Renderer(IServiceProvider serviceProvider, ILoggerFactory loggerFactory,
             _serviceProvider = serviceProvider;
             _logger = loggerFactory.CreateLogger<Renderer>();
             _componentFactory = new ComponentFactory(componentActivator);
+
+            _hotReloadEnvironment = serviceProvider.GetService<HotReloadEnvironment>() ?? HotReloadEnvironment.Instance;
```

Comment:
Might be nice to have a comment here stating why the DI service is optional. What does it mean to fall back on the singleton? Should the singleton be renamed to something like `HotReloadEnvironment.HotReloadDisabledInstance`?

Reply:
My guess is that the singleton is to keep the enabled-and-disabled code paths aligned enough that it's simple to configure the linker to strip it all out. Is that right? If so it seems like the singleton does represent the "disabled" case.